modules/core/shared/src/main/scala-3/io/circe/derivation/ConfiguredEncoder.scala

```diff
@@ -54,10 +54,18 @@ trait ConfiguredEncoder[A](using conf: Configuration) extends Encoder.AsObject[A
           JsonObject.singleton(constructorName, json)
 
 object ConfiguredEncoder:
-  inline final def derived[A](using conf: Configuration)(using inline mirror: Mirror.Of[A]): ConfiguredEncoder[A] =
+  inline private def derivedBase[A](inline autoRecurse: Boolean)(using conf: Configuration)(using
+    inline mirror: Mirror.Of[A]
+  ): ConfiguredEncoder[A] =
     new ConfiguredEncoder[A] with SumOrProduct:
       lazy val elemLabels: List[String] = summonLabels[mirror.MirroredElemLabels]
-      lazy val elemEncoders: List[Encoder[?]] = summonEncoders[mirror.MirroredElemTypes]
+      lazy val elemEncoders: List[Encoder[?]] =
+        summonEncoders[mirror.MirroredElemTypes](
+          autoRecurse,
+          inline mirror match
+            case _: Mirror.ProductOf[A] => false
+            case _: Mirror.SumOf[A]     => true
```

Comment:
I was hoping to share this logic with `isSum` below by making `isSum` an `inline val`, e.g.

```scala
inline val isSum =
  inline mirror match
    case _: Mirror.ProductOf[A] => false
    case _: Mirror.SumOf[A]     => true

lazy val elemEncoders: List[Encoder[?]] = summonEncoders[mirror.MirroredElemTypes](autoRecurse, isSum)
```

but then I get this error:

```scala
 79 |    given encoder: Encoder.AsObject[Adt1] = Encoder.AsObject.derivedNoAutoRecursion
    |                                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |Cannot reduce `inline if` because its condition is not a constant value: this.isSum
```

I'm sure there's something I'm misunderstanding about `inline val`s, just wanted to call it out